pkg/coordinator/deposits.go

```diff
@@ -67,81 +66,180 @@ func ListDeposits(
 	return nil
 }
 
-func removeSwept(deposits []depositEntry) []depositEntry {
-	result := []depositEntry{}
-	for _, deposit := range deposits {
-		if deposit.isSwept {
-			continue
+// FindDepositsToSweep finds deposits that can be swept.
+// If a wallet public key hash is provided, it will find unswept deposits for the
+// given wallet. If a wallet public key hash is nil, it will check all wallets
+// starting from the oldest one to find a first wallet containing unswept deposits
+// and return those deposits.
+// maxNumberOfDeposits is used as a ceiling for the number of deposits in the
+// result. If number of discovered deposits meets the maxNumberOfDeposits the
+// function will stop fetching more deposits.
+// This function will return a wallet public key hash and a list of deposits from
+// the wallet that can be swept.
+// Deposits with insufficient number of funding transaction confirmations will
+// not be taken into consideration for sweeping.
+// The result will not mix deposits for different wallets.
+// If no unswept deposits were found the function will return ErrNoDepositsToSweep
+// error.
+// TODO: Add unit tests
+// TODO: Cache immutable data
+// TODO: Don't call chain for old data
+func FindDepositsToSweep(
+	tbtcChain TbtcChain,
+	btcChain bitcoin.Chain,
+	walletPublicKeyHash *WalletPublicKeyHash,
+	maxNumberOfDeposits uint16,
+) (WalletPublicKeyHash, []*DepositSweepDetails, error) {
+	logger.Infof("deposit sweep max size: %d", maxNumberOfDeposits)
+
+	getDepositsToSweepFromWallet := func(walletToSweep WalletPublicKeyHash) ([]depositEntry, error) {
+		unsweptDeposits, err := getDeposits(
+			tbtcChain,
+			btcChain,
+			&walletToSweep,
+			int(maxNumberOfDeposits),
+			true,
+			true,
+		)
+		if err != nil {
+			return nil,
+				fmt.Errorf(
+					"failed to get deposits for [%s] wallet: [%w]",
+					walletToSweep,
+					err,
+				)
+		}
+		return unsweptDeposits, nil
+	}
+
+	depositsToSweep := make([]depositEntry, 0, maxNumberOfDeposits)
+	// If walletPublicKeyHash is not provided we need to find a wallet that has
+	// unswept deposits.
+	if walletPublicKeyHash == nil {
+		walletRegisteredEvents, err := tbtcChain.PastNewWalletRegisteredEvents(nil)
+		if err != nil {
+			return WalletPublicKeyHash{}, nil, fmt.Errorf("failed to get registered wallets: [%w]", err)
+		}
+
+		// Take the oldest first
+		sort.SliceStable(walletRegisteredEvents, func(i, j int) bool {
+			return walletRegisteredEvents[i].BlockNumber < walletRegisteredEvents[j].BlockNumber
+		})
+
+		// TODO: Optimization: Determine which wallets are already fully swept and
+		// cannot take any new deposits so we don't query them again.
```

Comment:
Only the two most recently created wallets in the system are sweeping. The most recent one is sweeping because it's an active wallet. The almost-most-recently created wallet is sweeping because it was the active wallet a short time ago and there is a chance some deposits were revealed at the moment the active wallet was switched or that the reveal transactions were in Ethereum mempool for a while.

That said, we should take only two of the most recently created wallets in this loop. I don't think we can have any other good optimization - checking which wallet is fully swept always requires the same algorithm to be executed: check all revealed deposits on Ethereum and see if they are all swept.